### ISSUE TITLE
Godot 4 And Above Freezing When Saving Files

### ISSUE THREAD

**Issue by MMT007** [godotengine/godot#90366]
### Tested versions

- Tested In Godot 4.0 stable and 4.2.1 Stable

### System information

Godot v4.2.1.stable - Ubuntu 22.04.4 LTS 22.04 - Wayland - GLES3 (Compatibility) - Mesa Intel(R) HD Graphics (BYT) () - Intel(R) Celeron(R) CPU  N2806  @ 1.60GHz (2 Threads)

### Issue description

Sometimes When I Try To Save A Scene, Or Creating A Script To A Node, Godot Just Randomly Freezes And Becomes Unusable. It Usually Happens When I Try To Save A Scene, But It Also Happened When I Tried Creating A Script Or Saving A Node As A Scene. ( Godot 3.6-beta4 and below work as intended )

[Screencast from 07-04-2024 18:38:56.webm](https://github.com/godotengine/godot/assets/67060689/ebfe58cd-0e8b-4dbf-9ce8-97744aad42bc)

### Steps to reproduce

- Open Godot 4.0 stable or Above
- Load A Project
- Save A Scene/Create A Script
- Bug Happens

### Minimal reproduction project (MRP)

N/A

### --verbose Output
[Godot Log.txt](https://github.com/godotengine/godot/files/14972400/Godot.Log.txt)

**Comment by berarma** [comment]
I'm seeing this issue too with Godot 4.2.2-rc3 and 4.3-dev5. I don't think it happened with 4.2.1, maybe it did but not that frequently. It happens randomly.

My system is Debian GNU/Linux 12.5, X11, Nvidia RTX 2700 Super with official drivers 525.147.05.

Godot becomes completely unresponsive, except that I can still switch from the dialog to the main window but it doesn't react to anything. I have to kill it from a terminal.

**Comment by MMT007** [comment]
> I'm seeing this issue too with Godot 4.2.2-rc3 and 4.3-dev5. I don't think it happened with 4.2.1, maybe it did but not that frequently. It happens randomly.
> 
> My system is Debian GNU/Linux 12.5, X11, Nvidia RTX 2700 Super with official drivers 525.147.05.
> 
> Godot becomes completely unresponsive, except that I can still switch from the dialog to the main window but it doesn't react to anything. I have to kill it from a terminal.


Me, along with some peoole on discord foud that using `--single-window` when opening godot, helps minimize the freezes

**Comment by berarma** [comment]
Thanks @MMT007.

@Sauermann What more testing can we do? Do you need some more info?

It seems to happen after the actual saving of files since I haven't lost any edits so far. It's a soft lock, a TERM signal forces quit.

**Comment by tufduckhotmailcom** [comment]
I get this issue as well in Godot 4.2.2. I open my project, do nothing else then just press Ctrl+S, and then Godot hangs. I have repeated this 3 times (killed the Godot process, reopened Godot, press Ctrl+S), and it hangs everytime.

This never happened to me on Godot 4.1.
I'm on Windows. 

Funny as I was typing this it suddenly unfroze, after about one minute. And now I can press Ctrl+S repeatedly and it doesn't freeze. Oh well.

**Comment by tufduckhotmailcom** [comment]
Ok, I must correct myself. I just downloaded Godot 4.1.4 and tried saving my project again. It immediately froze. So me saying it never happened in Godot 4.1 wasn't actually true. 

But my old game which is completed now, and which I developed all the time in 4.1, (maybe about 100 hours in total) never froze. Not even a single time. 

So it's something with my new project that's causing this. And it could be one of the addons I have installed. I will try to remove them and see is the issue goes away.

**Comment by MMUTFX2053** [comment]
i have this issue too, from time to time, not often, and then it hangs exactly on the same spot on the progress bar, around 70%

**Comment by MMUTFX2053** [comment]
maybe this is useful to note, for those that are going to fix the bug, if godot freezes, and i open it up again the changes that i made before the freeze did get saved

**Comment by MMUTFX2053** [comment]
took a screenshot, this is the exact spot where godot freezes for me every time, seems to be during thumbnail creation

![Screenshot from 2024-05-14 00-14-42](https://github.com/godotengine/godot/assets/144059118/f28eb573-82dc-4798-bbf5-2685daec76b4)

**Comment by Zireael07** [comment]
In that case likely a duplicate of #370974 which was just fixed by #91630

**Comment by Smorty10** [comment]
I am experiencing the same issue as op, where it freezes at 100 percent, but it does still save my progress. I am on the latest dev build (Godot 4.3 dev 6) but it has happened ever since Godot 4.

**Comment by MMUTFX2053** [comment]
@Smorty10 i think the fixes that @Zireael07 mentioned will fix the issue, but not 100% certain

**Comment by iconberg** [comment]
I have this error too when i run a scene and changes are saved.
Save windows hangs with 100% and i have to kill godot.

Godot v4.2.2.stable - Linux Mint 20 (Ulyana) - X11 - Vulkan (Mobile) - dedicated NVIDIA GeForce GTX 1660 Ti (nvidia; 535.171.04) - Intel(R) Core(TM)2 Quad  CPU   Q8200  @ 2.33GHz (4 Threads)

**Comment by why-try313** [comment]
@iconberg Can confirm that using ```--single-window``` as mentioned by MMT007 reduces not only save freezes but click issues (renaming and list selector), color picker lag and importer lag ```on X11```,  looks like a conflict with the desktop environment - To add a note, since Mint uses multiple desktop environments, I should specify I'm using `xfce4` so might be related if you're using it too.

> > I'm seeing this issue too with Godot 4.2.2-rc3 and 4.3-dev5. I don't think it happened with 4.2.1, maybe it did but not that frequently. It happens randomly.
> > My system is Debian GNU/Linux 12.5, X11, Nvidia RTX 2700 Super with official drivers 525.147.05.
> > Godot becomes completely unresponsive, except that I can still switch from the dialog to the main window but it doesn't react to anything. I have to kill it from a terminal.
> 
> Me, along with some peoole on discord found that using `--single-window` when opening godot, helps minimize the freezes

**Comment by QbieShay** [comment]
@MMUTFX2053 A friend of mine is experiencing this issue as well. Did you ever get to pause the process and look where it's stuck?

**Comment by MMUTFX2053** [comment]
@QbieShay not sure what you mean, this happens sometimes during saving, but when it does godot freezes up completely, nothing else i can do but to terminate it and restart, but i did notice that the changes are getting saved, also, deleting the cache seems to help a bit, the freezes happen a bit less often.

**Comment by QbieShay** [comment]
I was talking about attaching a debugger to Godot to see where it's stuck. I'll see if friend can repro

**Comment by MMUTFX2053** [comment]
im still getting this error, in 4.3 stable

**Comment by regulus79** [comment]
I am also experiencing random crashes when saving on 4.4dev. It doesn't happen most of the time, but occasionally it will freeze at 100%, and I have to kill the process to stop it.

**Comment by gattolfo98** [comment]
same here on 4.3stable, it block saving at 75% or 100%, my only solution is to kill app

**Comment by seetherdragon** [comment]
This happens to me VERY frequently (freezing during save with the thumbnail message) in 4.3 stable in Wayland Ubuntu 24 LTS (one monitor only).  I have to kill the editor app several times in a day.  Very irritating.

**Comment by oddlytimbot** [comment]
I teach a class in Godot, usually every student experiences this a couple of times per session with Godot 4.3 Stable. It is always upon saving, and always hangs at "creating thumbnail" when saving a scene. We are on Ubuntu 24.04.1 LTS. Kernel Version 6.8.0-49-generic. Window system is Wayland.

**Comment by berarma** [comment]
I don't think closing an issue because it no longer bothers you is the right way to go about a bug. It can still matter to others, and it does judging by the amount of reactions and subscribers. I guess it would be better letting it open and unsubscribe if you're no longer interested.

**Comment by rneeft** [comment]
Is there a way to disable creating Thumbnail? Godot engine almost becomes unusable,  when every 5 saves Ggodot hangs. Running System Monitor next to Godot to kill is more quicker.  Running Ubuntu 24.04 and latest Godot.

**Comment by oddlytimbot** [comment]
> Is there a way to disable creating Thumbnail? Godot engine almost becomes unusable, when every 5 saves Ggodot hangs. Running System Monitor next to Godot to kill is more quicker. Running Ubuntu 24.04 and latest Godot.

This happened to me four times while teaching a class, along with half the students. I have a suspicion this is a windowing problem rather than the actual thumbnail creation, because both the save and the thumbnail seem to be completed despite it hanging. That really makes me think it is a Wayland problem affecting Ubuntu in particular since that OS defaults to Wayland on install now. I have had a problem with another software I use and was able to work around it by launching with a flag that set the GDK to use X11. Does anyone know if any flags like this exist when launching Godot?

**Comment by berarma** [comment]
> > Is there a way to disable creating Thumbnail? Godot engine almost becomes unusable, when every 5 saves Ggodot hangs. Running System Monitor next to Godot to kill is more quicker. Running Ubuntu 24.04 and latest Godot.
> 
> This happened to me four times while teaching a class, along with half the students. I have a suspicion this is a windowing problem rather than the actual thumbnail creation, because both the save and the thumbnail seem to be completed despite it hanging. That really makes me think it is a Wayland problem affecting Ubuntu in particular since that OS defaults to Wayland on install now. I have had a problem with another software I use and was able to work around it by launching with a flag that set the GDK to use X11. Does anyone know if any flags like this exist when launching Godot?

It's happened to me, and I don't use Wayland. There have been days when it has happened frequently, but now it hasn't happened for several weeks. Since Godot has some issues related to the editor cache and I've had to delete it several times, I think it might be related to the editor cache, too. I'd try deleting the cache, although this won't prevent it from happening again but might mitigate it.

**Comment by Zireael07** [comment]
> I have a suspicion this is a windowing problem rather than the actual thumbnail creation, because both the save and the thumbnail seem to be completed despite it hanging

Agreed

**Comment by oddlytimbot** [comment]
> It's happened to me, and I don't use Wayland. There have been days when it has happened frequently, but now it hasn't happened for several weeks. Since Godot has some issues related to the editor cache and I've had to delete it several times, I think it might be related to the editor cache, too. I'd try deleting the cache, although this won't prevent it from happening again but might mitigate it.

Same here - last week it didn't happen to me at all, and only once to a student over a three hour session. This week, it happened continually. I'll try clearing the editor cache, thanks for the suggestion.

**Comment by astrale-sharp** [comment]
Same here, very frequently, I'm surprised this issue has so little attention, do we have clues what's causing it?

**Comment by pennyloafers** [comment]
I recently upgraded to 24.04 and started experiencing these issues. 
I think its getting stuck on closing the popup window. 
here is  a back trace, I don't have the debug symbols but I will try to reproduce with them next time.
```
#0  0x0000783d0d71b4cd in __GI___poll (fds=0x7fff9a53c908, nfds=1, timeout=-1) at ../sysdeps/unix/sysv/linux/poll.c:29
#1  0x0000783d0bbe38ca in ?? () from /lib/x86_64-linux-gnu/libxcb.so.1
#2  0x0000783d0bbe528c in xcb_wait_for_event () from /lib/x86_64-linux-gnu/libxcb.so.1
#3  0x0000783d0c447258 in _XReadEvents () from /lib/x86_64-linux-gnu/libX11.so.6
#4  0x0000783d0c42c380 in XIfEvent () from /lib/x86_64-linux-gnu/libX11.so.6
#5  0x0000783d0c4747ab in ?? () from /lib/x86_64-linux-gnu/libX11.so.6
#6  0x0000783d0c475e95 in ?? () from /lib/x86_64-linux-gnu/libX11.so.6
#7  0x0000783d0c479b3a in _XimRead () from /lib/x86_64-linux-gnu/libX11.so.6
#8  0x0000783d0c4641aa in ?? () from /lib/x86_64-linux-gnu/libX11.so.6
#9  0x0000783d0c44f967 in XDestroyIC () from /lib/x86_64-linux-gnu/libX11.so.6
#10 0x00000000004b2e3a in ?? ()
#11 0x00000000023dc7de in ?? ()
```
ubuntu 24.04 defaults to wayland and according to the source only X11 display server calls XDestroyIC. I wonder if there is a display server selection issue as there should be a wayland display server.

I will try to play around with this some more, but my experience is limited.

**Comment by pennyloafers** [comment]
A little digging around the source code, there is a setup in main.cpp

```
/* Initialize Display Server */

	{
		OS::get_singleton()->benchmark_begin_measure("Servers", "Display");

		if (display_driver.is_empty()) {
			display_driver = GLOBAL_GET("display/display_server/driver");
		}

		int display_driver_idx = -1;

		if (display_driver.is_empty() || display_driver == "default") {
			display_driver_idx = 0;
		} else {
			for (int i = 0; i < DisplayServer::get_create_function_count(); i++) {
				String name = DisplayServer::get_create_function_name(i);
				if (display_driver == name) {
					display_driver_idx = i;
					break;
				}
			}
```

if the project settings (GLOBAL_GET("display/display_server/driver")) is not changed from "default" the `diplay_driver_idx` will equal  0. Which is X11 by default because of the registration order.

```
#ifdef X11_ENABLED
	DisplayServerX11::register_x11_driver(); // driver index 0
#endif

#ifdef WAYLAND_ENABLED
	DisplayServerWayland::register_wayland_driver();
#endif
``` 
This is why the X11 display driver is being used even when wayland is the current OS display server.

**Comment by oddlytimbot** [comment]
Really appreciate this research @pennyloafers. I will try explicitly setting the display server and will get my students to do the same. We should know pretty quickly if this fixes things.

**Comment by pennyloafers** [comment]
I found that you can simply reproduce the issue if you spam saving (`ctrl + s`).

@oddlytimbot, it is actually better to start the editor from the command line like so `godot --display-driver wayland` as the editor doesn't have a project settings you can change. I think project settings is just for the running game window.

If you use Ubuntu you can actually change the OS display server to X11 before login. 

![Image](https://github.com/user-attachments/assets/d33632de-0943-480d-8cdf-06cc6c853c71)

you can also check what the Ubuntu is running with this command. `echo $XDG_SESSION_TYPE` which will print x11 or wayland.

Anyway, running Godot with the wayland driver seems to have bugs, as the window title bar is missing and its hard to close out of windows without it.

But, i did find that running with wayland on wayland, OR, x11 on x11, seemed to prevent the popup freeze issue.

**Comment by akien-mga** [comment]
You can configure the editor on Linux to use the Wayland DisplayServer with this editor setting: `run/platforms/linuxbsd/prefer_wayland`

It's currently not the default (so we intentionally default to using XWayland on Wayland) because the Wayland DisplayServer doesn't have yet feature parity with the X11 one. But indeed in some cases where X11/XWayland is buggy it might be a better experience already.

**Comment by akien-mga** [comment]
There seems to be a bit of mixed reports here for multiple Godot versions and OSes, but the original and most recent ones seem to be specifically about Ubuntu + Wayland, so I'll rescope this issue to focus on that only.

If anyone experiences such freeze when saving files on Windows, please open a dedicated issue, as such issues are almost guaranteed to be platform specific (so what Ubuntu users experience is not the same bug that Windows users might experience, even if the symptoms - a freeze - might be similar).

**Comment by berarma** [comment]
I don't think this is specific to any distro or windowing system. I'll restate that I'm having these freezes on Debian with X11 although not as frequently as before, and in my experience it might be somewhat related to the editor cache.

The new subject is confusing, what's "Ubuntu  Wayland with X11 DisplayServer"? You either are running one or the other. Or maybe the X11 compatibility layer over Wayland. But like I said I don't think this is a Wayland issue.

**Comment by akien-mga** [comment]
> The new subject is confusing, what's "Ubuntu Wayland with X11 DisplayServer"? You either are running one or the other. Or maybe the X11 compatibility layer over Wayland.

"X11 DisplayServer" refers to Godot's own DisplayServerX11, which is using Xlib. That's what is used by default unless you explicitly configure things to use Godot's DisplayServerWayland.

So even on a Wayland *session* on Linux, it will be using Xlib via XWayland.

**Comment by fabrb** [comment]
Bumping this. I tested two things:
1. Fiddle with Godot with only the built-in display
2. Fiddle with Godot after connecting and settings the display configuration to "Join Displays"

The first case worked without a single problem, the engine and other applications on my machine did not freeze at all.
When I plugged my second monitor, it immediately froze my machine for 1~2 seconds after I clicked to restart the project button.

Since I am quite new to Linux, I can not tell for how long this problem is occurring.

My specs:
Hardware: Acer Nitro ANV15-51
Memory: 16.0 GiB
Processor: 13th Gen Intel® Core™ i7-13620H × 16
Graphics: NVIDIA Corporation / Mesa Intel® Graphics (RPL-P)
OS: Pop!_OS 22.04 LTS
GNOME: GNOME Shell 42.9

**Comment by mcdz** [comment]
Check the scene file size please. I haven't read all the comments, but this happened to me once, something to do with theme in my case.

**Comment by Guuri11** [comment]
Same here, Ubuntu 22, Godot 4.4.1, it freezes every time I try to save, so frustrating

**Comment by pennyloafers** [comment]
> Same here, Ubuntu 22, Godot 4.4.1, it freezes every time I try to save, so frustrating

Ive updated to ubuntu 24, and still use x11 and i havent run into the issue in a long time. Maybe never in 24.
